File: dp/0072-minDistance.py
"""
"""
class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)
        @cache
        def dfs(i, j):
            if i < 0: return j+1;
            if j < 0: return i+1;
            if word1[i] == word2[j]: return dfs(i-1, j-1)
            return min(dfs(i-1, j), dfs(i, j-1), dfs(i-1, j-1))+1
        return dfs(m-1, n-1) 

class Solution:
    def minDistance(self, word1: str, word2: str) -> int:
        m, n = len(word1), len(word2)
        f = [[0]*(n+1) for _ in range(m+1)]
        f[0] = list(range(n+1))
        for i, x enumerate(word1):
            f[i+1][0] = i+1
            for j, y in enumerate(word2):
                f[i+1][j+1] = f[i][j] if x == y else min(f[i][j+1], f[i+1][j], f[i][j])+1
        return f(m, n)   

File: dp/0072-minDistance.cpp
/*
 * leetcode 72
 * min distance
 *
 * three operations:
 * - insert a char
 * - delete a char
 * - replace a char
 *
 */

class Solution {
public:
    int minDistance(string word1, string word2) {
        int dp[501][501];
        // dp[i][j] definition
        // the distance from first i chars in word1 
        // to first j chars in word2

        int n = word1.size();
        int m = word2.size();
        // initialize
        for (int i = 0; i <= n; i++) dp[i][0] = i;
        for (int j = 0; i <= m; j++) dp[0][j] = j;
        // dp state transfer
        // add: dp[i][j] = dp[i][j-1] + 1 
        // del: dp[i][j] = dp[i-1][j] + 1 
        // replace: dp[i][j] = dp[i-1][j-1] + 1
        for (int i = 1; i <= n; i++) {
            for (int j = 1; j <= m; j++) {
                if (word1[i-1] == word2[j-1])
                    dp[i][j] = dp[i-1][j-1];
                else {
                    dp[i][j] = 1 + min(dp[i-1][j-1], dp[i-1][j], dp[i][j-1]);
                }        
            }
        } 
        return dp[n][m];
    }
};

class Solution {
public:
    int minDistance(string word1, string word2) {
        int m = word1.size();
        int n = word2.size();
        vector<vector<int> > dp(m+1, vector<int>(n+1, 0));
        // initialize
        for (int i = 0; i <= m; i++) {
            dp[i][0] = i;
        }
        for (int j = 0; j <= n; j++) {
            dp[0][j] = j;
        }
        // state transfer
        for (int i = 1; i <= m; i++) {
            for (int j = 1; j <= n; j++) {
                if (word1[i-1] == word2[j-1]) {
                    dp[i][j] = dp[i-1][j-1];
                } else {
                    dp[i][j] = 1 + min(dp[i-1][j-1], min(dp[i-1][j], dp[i][j-1]));
                }
            }
        }
        return dp[m][n];
        // dp.back().back();
    }
};

